routing.go

```diff
@@ -441,6 +441,23 @@ func (dht *IpfsDHT) Provide(ctx context.Context, key cid.Cid, brdcst bool) (err
 	return ctx.Err()
 }
 
+func (dht *IpfsDHT) OptimisticProvide(ctx context.Context, key cid.Cid) (err error) {
```

Comment:
I would not expose this as a second method, now every caller needs to have it's `if` check to know what to call.

I would do `if dht.enableOptProv {` in the `Provide` method, and then dispatch to either classical or optimistic client.
I do not think we are gonna start having provides where some of them can be done optimistically and other where we really want them to be precise.

Reply:
Removed public API and added the following to the `Provide` method:

```go
	if dht.enableOptProv {
		err := dht.GetAndProvideToClosestPeers(ctx, string(keyMH))
		if errors.Is(err, netsize.ErrNotEnoughData) {
			logger.Debugln("not enough data for optimistic provide taking classic approach")
		} else {
			return err
		}
	}
```

I thought of falling back to the classic approach if we don't have enough data points for the optimistic one.

Reply:
@Jorropo IMO a second method makes sense as `OptimisticProvide` isn't exactly the same use case as the `Provide` operation (see [this comment](https://github.com/libp2p/go-libp2p-kad-dht/pull/783#pullrequestreview-1347928916)). WDYT?

Reply:
@guillaumemichel will we write code anytime soon that is Optimistic Provide aware ?

By that I mean will we have code that use different clients for different workloads ?

I don't belive so, thus no need to make the API more complex and require all consumers to purposely call the right method.

Reply:
> will we write code anytime soon that is Optimistic Provide aware ?
I don't think so, as IMO kubo isn't the best use case for Optimistic Provide.

However, if other applications want to use it (the goal of this feature is still to be useful), let's not screw their reprovide operation, otherwise no one will want to use this feature at all.

IMO the job achieved by optimistic provide is different as the one achieved by the normal Provide, and each have their use cases, let users use both.